crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -156,143 +184,78 @@ where
-            // - when indexer is up to date, we download at least one checkpoint at a time.
-            current_parallel_downloads =
-                std::cmp::min(downloaded_checkpoints.len() + 1, MAX_PARALLEL_DOWNLOADS);
-            if downloaded_checkpoints.is_empty() {
-                warn!(
-                    "No checkpoints were downloaded for sequence number {}, retrying...",
-                    next_cursor_sequence_number
-                );
-                continue;
-            }
-
-            // Index checkpoint data
-            let index_timer = self.metrics.checkpoint_index_latency.start_timer();
-            let indexed_checkpoint_epoch_vec = join_all(downloaded_checkpoints.iter().map(
-                |downloaded_checkpoint| async {
-                    self.index_checkpoint_and_epoch(downloaded_checkpoint).await
-                },
-            ))
-            .await
-            .into_iter()
-            .collect::<Result<Vec<_>, IndexerError>>()
-            .map_err(|e| {
-                error!(
-                    "Failed to index checkpoints {:?} with error: {}",
-                    downloaded_checkpoints,
-                    e.to_string()
-                );
-                e
-            })?;
-            let (indexed_checkpoints, indexed_epochs): (Vec<_>, Vec<_>) =
-                indexed_checkpoint_epoch_vec.into_iter().unzip();
-            index_timer.stop_and_record();
+    async fn start_checkpoint_commit(&self) -> Result<(), IndexerError> {
+        info!("Indexer checkpoint commit task started...");
+        let checkpoint_commit_batch_size = env::var("CHECKPOINT_COMMIT_BATCH_SIZE")
+            .unwrap_or(5.to_string())
+            .parse::<u64>()
+            .unwrap();
+        loop {
+            let mut checkpoint_receiver_guard = self.checkpoint_receiver.lock().await;
 
-            for epoch in indexed_epochs.into_iter().flatten() {
-                // commit first epoch immediately, send other epochs to channel to be committed later.
-                if epoch.last_epoch.is_none() {
-                    let epoch_db_guard = self.metrics.epoch_db_commit_latency.start_timer();
-                    info!("Persisting first epoch...");
-                    let mut persist_first_epoch_res = self.state.persist_epoch(&epoch).await;
-                    while persist_first_epoch_res.is_err() {
-                        warn!("Failed to persist first epoch, retrying...");
-                        persist_first_epoch_res = self.state.persist_epoch(&epoch).await;
+            let mut additional_checkpoints = checkpoint_commit_batch_size;
+            let mut indexed_checkpoint_batch = vec![];
+            loop {
+                if let Ok(ckp) = checkpoint_receiver_guard.try_recv() {
+                    indexed_checkpoint_batch.push(ckp);
+                    additional_checkpoints -= 1;
+                    if additional_checkpoints == 0 {
```

Comment:
during the last batch before caught-up, for example if the batch size is 10 and we only have 5 in the last batch, looks like we will be waiting for another 5 checkpoints to be generated and then downloaded before committing?

this is prob fine if it's only about the last batch before caught-up, it could be worse if a new checkpoint is generated while committing these 10 checkpoints, b/c in that sense, here we will try to wait for 10 checkpoints before committing the next batch.

Reply:
resolved offline